beginner_source/blitz/autograd_tutorial.py

```diff
@@ -1,198 +1,276 @@
+the parameters using **gradient descent**. For a more detailed walkthrough
+of backprop, check out this `video from
+3Blue1Brown <https://www.youtube.com/watch?v=tIeHLnjs5U8>`__.
+
+Most deep learning frameworks use automatic differentiation for
+backprop; in PyTorch, it is handled by Autograd.
+
+
+Usage in PyTorch
+~~~~~~~~~~~
+Backward propagation can be kicked off by calling ``.backward()`` on the error tensor.
+This collects the gradients for each parameter in the model.
+"""
 
+import torch, torchvision
 
-The ``autograd`` package provides automatic differentiation for all operations
-on Tensors. It is a define-by-run framework, which means that your backprop is
-defined by how your code is run, and that every single iteration can be
-different.
+model = torchvision.models.resnet18(pretrained=True)
+data = torch.rand(1, 3, 64, 64)
+labels = torch.rand(1, 1000)
+prediction = model(data) # forward pass
+loss = (prediction - labels).sum()
+loss.backward() # backward pass
 
-Let us see this in more simple terms with some examples.
+optim = torch.optim.SGD(model.parameters(), lr=1e-2, momentum=0.9)
+optim.step() #gradient descent
 
-Tensor
---------
+######################################################################
+# At this point, you have everything you need to build your neural network.
+# The below sections detail the workings of autograd - feel free to skip them.
+#
 
-``torch.Tensor`` is the central class of the package. If you set its attribute
-``.requires_grad`` as ``True``, it starts to track all operations on it. When
-you finish your computation you can call ``.backward()`` and have all the
-gradients computed automatically. The gradient for this tensor will be
-accumulated into ``.grad`` attribute.
 
-To stop a tensor from tracking history, you can call ``.detach()`` to detach
-it from the computation history, and to prevent future computation from being
-tracked.
+######################################################################
+# --------------
+#
 
-To prevent tracking history (and using memory), you can also wrap the code block
-in ``with torch.no_grad():``. This can be particularly helpful when evaluating a
-model because the model may have trainable parameters with
-``requires_grad=True``, but for which we don't need the gradients.
 
-There’s one more class which is very important for autograd
-implementation - a ``Function``.
+######################################################################
+# Differentiation in Autograd
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-``Tensor`` and ``Function`` are interconnected and build up an acyclic
-graph, that encodes a complete history of computation. Each tensor has
-a ``.grad_fn`` attribute that references a ``Function`` that has created
-the ``Tensor`` (except for Tensors created by the user - their
-``grad_fn is None``).
+# The ``requires_grad`` flag lets autograd know
+# if we need gradients w.r.t. these tensors. If it is ``True``, autograd
```

Comment:
This doesn't appear to render correctly as text in the preview: https://deploy-preview-1185--pytorch-tutorials-preview.netlify.app/beginner/blitz/autograd_tutorial.html#differentiation-in-autograd